### PR TITLE
Save results of A/B test runs in a JSON file for future processing

### PR DESCRIPTION
**Pull request by flar**
This is a proof of concept for writing the results of an A/B run to a JSON file.

I also modified the `ab.dart` source file to be a `main()` program that can read the files back in and print out reports or do other processing. Thus far, the only things it can do is to write out the same reports that were originally printed during the test, but more processing could be added in the future, such as combining runs, generating some code for graphs, etc.

My naming conventions probably violate a number of dart/json style and naming guidelines, but suggestions are welcome!

The PR is marked WIP, because I know it will have to be revised quite a bit before it goes back, but I do specifically want some input at this point so ignore the "do not review" part of the label...

### CODE REVIEW

File: dev/devicelab/bin/run.dart [hunk 1]
```diff
@@ -245,6 +266,10 @@ final ArgParser _argParser = ArgParser()
       }
     },
   )
+  ..addOption(
+    'abresultfile',
```
Comment (yjbanov) [inline dev/devicelab/bin/run.dart:270]: nit: other options in our tools use the dash to separate words, e.g. `ab-result-file`.

File: dev/devicelab/lib/framework/ab.dart [hunk 2]
```diff
@@ -2,23 +2,102 @@
...
+const String kBenchmarkABVersion = '1.0';
+
+void _usage(String error) {
+  stderr.writeln(error);
+  stderr.writeln('Usage:\n');
+  stderr.writeln(_argParser.usage);
+  exitCode = 1;
+}
+
+Future<void> main(List<String> rawArgs) async {
```
Comment (yjbanov) [inline dev/devicelab/lib/framework/ab.dart:31]: Would it be possible to move the `main` under `bin/`? The Dart VM doesn't like entrypoints under `lib/`. It's also a pub convention (e.g. you can `pub global activate` an executable from `bin/`).

File: dev/devicelab/lib/framework/ab.dart [hunk 3]
```diff
@@ -2,23 +2,102 @@
...
   ///
   /// The result may contain multiple score keys.
   ///
   /// [result] is expected to be a serialization of [TaskResult].
   void addAResult(Map<String, dynamic> result) {
-    _addResult(result, _aResults);
+    if (_runEnd == null) {
+      _addResult(result, _aResults);
+    } else {
+      throw StateError('Cannot add results to ABTest after it is finalized');
```
Comment (yjbanov) [inline dev/devicelab/lib/framework/ab.dart:99]: nit: I like checking for bad state at the beginning of the function. This way you throw early and you don't need to indent the normal logic.

File: dev/devicelab/lib/framework/ab.dart [hunk 4]
```diff
@@ -27,9 +106,28 @@ class ABTest {
   ///
   /// [result] is expected to be a serialization of [TaskResult].
   void addBResult(Map<String, dynamic> result) {
-    _addResult(result, _bResults);
+    if (_runEnd == null) {
+      _addResult(result, _bResults);
+    } else {
+      throw StateError('Cannot add results to ABTest after it is finalized');
```
Comment (yjbanov) [inline dev/devicelab/lib/framework/ab.dart:112]: same nit

File: dev/devicelab/lib/framework/ab.dart (hunk 2 above)
Comment (yjbanov) [reply dev/devicelab/lib/framework/ab.dart:31]: Would also be super-useful to have some docs in the README.md about how to use this.

File: dev/devicelab/bin/run.dart (hunk 1 above)
Comment (flar) [reply dev/devicelab/bin/run.dart:270]: Done

File: dev/devicelab/lib/framework/ab.dart (hunk 2 above)
Comment (flar) [reply dev/devicelab/lib/framework/ab.dart:31]: I'll do that. Also, I print the name of the output file at the end of a run - should I mention the "processing" program there too?

File: dev/devicelab/lib/framework/ab.dart (hunk 3 above)
Comment (flar) [reply dev/devicelab/lib/framework/ab.dart:99]: Done and...

File: dev/devicelab/lib/framework/ab.dart (hunk 4 above)
Comment (flar) [reply dev/devicelab/lib/framework/ab.dart:112]: ...and done...

**Comment by flar** [comment]
Some other naming considerations that I was looking for feedback on:

- default name of output file (currently `ABresults#.json`) (I also clarified the # in the options description)
- key names in JSON output (I didn't see any style guidelines on this and I just noticed that `benchmarkScoreKeys` uses camelCase)
- Name of method `ABTest.finalize()` (finalize, in retrospect, is also the name of a memory management mechanism in Java...)

**Comment by flar** [comment]
It would also be nice to save the frame times themselves so we can do more than just compare the summaries, but also graph the frame times against each other. That may be for a future PR...

File: dev/devicelab/lib/framework/ab.dart (hunk 2 above)
Comment (flar) [reply dev/devicelab/lib/framework/ab.dart]: tool code moved to `bin/summarize.dart`
`README.md` updated

**Comment by flar** [comment]
This is pretty much in its final form - wondering if we want to adjust the defaults for which kind of output appears in the terminal when an A/B test is run and which is the default output for the summarize tool.

(Also, more naming feedback would be nice.)

**Review by yjbanov — APPROVED** [review]

**Pull request merged**